Client/Frontend/UserContent/UserScripts/MainFrame/AtDocumentStart/TrackingProtectionStats.js

```diff
@@ -0,0 +1,136 @@
+
+  function onLoadNativeCallback() {
+    var messageHandler = window.webkit.messageHandlers.focusTrackingProtection;
+    var sendMessage = function(url) { messageHandler.postMessage({ url: url }) };
+
+    // Send back the sources of every script and image in the dom back to the host applicaiton
+    Array.prototype.map.call(document.scripts, function(t) { return t.src }).forEach(sendMessage);
+    Array.prototype.map.call(document.images, function(t) { return t.src }).forEach(sendMessage);
+  }
+
+  let originalOpen = null;
+  let originalSend = null;
+  let originalImageSrc = null;  
+  let mutationObserver = null;
+
+  function injectStatsTracking(enabled) {
+    if (enabled) {
+      if (originalOpen != null) {
+        return;
+      }
+      window.addEventListener("load", onLoadNativeCallback, false);
+    } else {
+      window.removeEventListener("load", onLoadNativeCallback, false);
+
+      if (originalOpen != null) { // if one is set, then all the enable code has run
+        XMLHttpRequest.prototype.open = originalOpen;
+        XMLHttpRequest.prototype.send = originalSend;
+        Image.prototype.src = originalImageSrc;
+        mutationObserver.disconnect();
+
+        originalOpen = originalSend = originalImageSrc = mutationObserver = null;
+      }
+      return;
+    }
+
+    var messageHandler = window.webkit.messageHandlers.focusTrackingProtection
+
+    // -------------------------------------------------
+    // Send ajax requests URLs to the host application
+    // -------------------------------------------------
+    var xhrProto = XMLHttpRequest.prototype;
+    if (originalOpen == null) {
+      originalOpen = xhrProto.open;
+      originalSend = xhrProto.send;
+    }
+
+    xhrProto.open = function(method, url) {
+      this._url = url;
+      return originalOpen.apply(this, arguments);
+    };
+
+    xhrProto.send = function(body) {
+      messageHandler.postMessage({
+        url: this._url
+      })
+      return originalSend.apply(this, arguments)
+    };
+
+    // -------------------------------------------------
+    // Detect when new sources get set on Image and send them to the host application
+    // -------------------------------------------------
+    if (originalImageSrc == null) {
+      originalImageSrc = Object.getOwnPropertyDescriptor(Image.prototype, 'src');
+    }
+    delete Image.prototype.src;
+    Object.defineProperty(Image.prototype, 'src', {
+      get: function() {
+        return originalImageSrc.get.call(this);
+      },
+      set: function(value) {
+        messageHandler.postMessage({
+          url: value
+        })
+        originalImageSrc.set.call(this, value);
+      }
+    });
+
+    // -------------------------------------------------
+    // Listen to when new <script> elements get added to the dom
+    // and send the source to the host application
+    // -------------------------------------------------
+    mutationObserver = new MutationObserver(function(mutations) {
```

Comment:
A couple things here:
1. I'd like to see if we can just hijack `HTMLScriptElement.prototype.src` like we do for `Image` above?
2. Does hijacking `Image.prototype.src` above actually work for appending `<img>`s to the DOM in all cases? i.e.: `var img = document.createElement('img'); img.src = '/foo.jpg'; document.body.appendChild(img);`
3. If we do actually have to use a `MutationObserver` for `<script>` elements, we need to at least check for the existence of `src` before calling `postMessage()`. For instance, I think it's possible to append a `<script>` to the DOM that has inline JS.

Reply:
You can test those cases with this JSBin: http://output.jsbin.com/rahizoq